File: 3969-valid-subarrays-with-matching-sum-digits-i/3969-valid-subarrays-with-matching-sum-digits-i.java
class Solution {
    public int countValidSubarrays(int[] nums, int x) {
        
         int n = nums.length;
        long[] prefix = new long[n + 1];

        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        int count = 0;
         for (int l = 0; l < n; l++) {
            for (int r = l; r < n; r++) {
                long sum = prefix[r + 1] - prefix[l];

                if (sum % 10 != x) continue;

                long firstDigit = sum;
                while (firstDigit >= 10) {
                    firstDigit /= 10;
                }
                 if (firstDigit == x) {
                    count++;
                }
            }
        }

        return count;

    }
}

File: 3969-valid-subarrays-with-matching-sum-digits-i/3969-valid-subarrays-with-matching-sum-digits-i.cpp
class Solution {
public:
    int countValidSubarrays(vector<int>& nums, int x) {
        int n = nums.size();
        int count = 0;
        for (int l = 0; l < n; l++) {
            long long sum = 0;
            for (int r = l; r < n; r++) {
                sum += nums[r];
                int last = sum % 10;
                long long tmp = sum;
                while (tmp >= 10) tmp /= 10;
                int first = (int)tmp;
                if (first == x && last == x) count++;
            }
        }
        return count;
    }
};